libflux/src/semantic/nodes.rs

```diff
@@ -229,37 +251,57 @@ pub struct File {
-                    let (env, cons) = stmt.infer(env, f)?;
-                    Ok((env, cons + rest))
-                }
-                Statement::Option(stmt) => {
-                    let (env, cons) = stmt.infer(env, f)?;
-                    Ok((env, cons + rest))
-                }
-                Statement::Expr(stmt) => {
-                    let (env, cons) = stmt.infer(env, f)?;
-                    Ok((env, cons + rest))
-                }
-                Statement::Test(stmt) => {
-                    let (env, cons) = stmt.infer(env, f)?;
-                    Ok((env, cons + rest))
-                }
-                Statement::Return(_) => Err(Error::invalid_statement(String::from(
-                    "cannot have return statement in file block",
-                ))),
-            },
-        )
-        // TODO: remove imported names from the type environment
+    fn infer(&mut self, mut env: Environment, f: &mut Fresher, i: &Importer) -> Result {
+        let mut imports = Vec::with_capacity(self.imports.len());
+
+        for dec in &self.imports {
+            let path = &dec.path.value;
+            let name = pkg_name_from_path(&path);
+
+            imports.push(name);
+
+            match i.import(path) {
+                Some(poly) => env.add(name.to_owned(), poly.clone()),
+                None => return Err(Error::unknown_import_path(path)),
+            };
+        }
+
+        let (mut env, constraints) =
+            self.body
+                .iter_mut()
+                .try_fold(
+                    (env, Constraints::empty()),
+                    |(env, rest), node| match node {
+                        Statement::Builtin(stmt) => {
+                            let (env, cons) = stmt.infer(env, f)?;
+                            Ok((env, cons + rest))
+                        }
+                        Statement::Variable(stmt) => {
+                            let (env, cons) = stmt.infer(env, f)?;
+                            Ok((env, cons + rest))
+                        }
+                        Statement::Option(stmt) => {
+                            let (env, cons) = stmt.infer(env, f)?;
+                            Ok((env, cons + rest))
+                        }
+                        Statement::Expr(stmt) => {
+                            let (env, cons) = stmt.infer(env, f)?;
+                            Ok((env, cons + rest))
+                        }
+                        Statement::Test(stmt) => {
+                            let (env, cons) = stmt.infer(env, f)?;
+                            Ok((env, cons + rest))
+                        }
+                        Statement::Return(_) => Err(Error::invalid_statement(String::from(
+                            "cannot have return statement in file block",
+                        ))),
+                    },
+                )?;
+
+        for name in imports {
+            env.remove(name);
```

Comment:
Is my understanding correct that it is safe to remove the names here because we have already recursed into all children of the file and so the env names are no longer needed?

Reply:
Yes correct. We do not want these names to stay in the environment as they are not visible to other files in the same package.